File: credit.py
from cs50 import get_int
import re

# prompt for input, check value type

while True:
    try:
        x = int(get_int("Number: "))
    except ValueError:
        continue
    if x < 0:
        continue
    else:
        break

# make a copy
copy = x

# find length
le = len(str(x))

# floor division: //
if le == 13 or le == 15 or le == 16:
    sum1 = 0
    sum2 = 0
    mod1 = 0
    mod2 = 0
    while x != 0:
        mod1 = int(x % 10)
        sum1 += mod1
        x //= 10
        mod2 = x % 10
        if mod2 * 2 > 9:
            sum2 += int(mod2 * 2 / 10 + mod2 * 2 % 10)
            x //= 10
        else:
            sum2 += mod2 * 2
            x //= 10
    
    # retrieve input
    x = str(copy)
    checksum = str(sum1 + sum2)

    if re.findall("0$", checksum):
        if le == 15:
            if re.findall("^34", x) or re.findall("^37", x):
                print("AMEX")
            else:
                print("INVAL6D")
        elif le == 13:
            if re.finall("^4", x):
                print("VISA")
            else:
                print("INVAL5D")
        elif le == 16:
            if re.findall("^4", x):
                print("VISA")
            elif re.findall("^51|52|53|54|55", x):
                # ^51|52|53|54|55 works
                # ^51|52|53|54|55 won't work
                print("MASTERCARD")
            else:
                print("INVALID")
        else:
            print("INVALID")
    else:
        print("INVALID")
else:
    print("INVALID")

File: week1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <ctype.h>
#include <math.h>

int main(void)
{
    long num;

    do
    {
        num = get_long("Number: ");
    }
    while (num < 0); // min valid num of digits is 13

    int x = 0;

    // wrong if int
    long m = num; // for finding x
    long n = num; // for checksum
    int mod1, mod2;
    int sum1 = 0;
    int sum2 = 0;
    // int checksum = sum1 + sum2;

    while (m) // m >= 0
    {
        m /= 10;
        x++;
    }

    if (x != 13 && x != 15 && x != 16)
    {
        printf("INVALID\n");

    }
    else
    {
        int i = 0;
        while (n != 0)
        {
            mod1 = n % 10;
            sum1 += mod1;
            n /= 10;

            mod2 = n % 10;

            if (mod2 * 2 > 9)
            {
                sum2 += mod2 * 2 / 10 + mod2 * 2 % 10 ;
                n /= 10;
            }
            else
            {
                sum2 += (mod2 * 2);
                n /= 10;

            }

        }

        int checksum = sum1 + sum2;

        //  TEST
        //  printf("sum1 %d\n", sum1); printf("sum2 %d\n", sum2); printf("checksum %d x %d \n", checksum, x);

        // checksum
        if (checksum % 10 == 0)
        {
            if (x == 15)
            {
                if (num / 10000000000000 == 37 || num / 10000000000000 == 34) // why pow(10, 15) not working
                {
                    printf("AMEX\n");
                }
                else
                {
                    printf("INVALID\n");

                }

            }
            else if (x == 13)
            {
                if (num / 1000000000000 == 4)
                {
                    printf("VISA\n");

                }
                else
                {
                    printf("INVALID\n");

                }
            }
            else if (x == 16)
            {
                if (num / 1000000000000000 == 4)
                {
                    printf("VISA\n");
                }
                else if (num / 100000000000000 == 51
                         || num / 100000000000000 == 52
                         || num / 100000000000000 == 53
                         || num / 100000000000000 == 54
                         || num / 100000000000000 == 55)
                {
                    printf("MASTERCARD\n");

                }
                else
                {
                    printf("INVALID\n");

                }

            }
            else
            {
                printf("INVALID\n");
            }

        }
        else
        {
            printf("INVALID\n"); // expect eof if wo
        }

    }

}